File: Data Structures & Algorithms/count-connected-components/submission-2.cpp
class Solution {
    vector<vector<int>> adj;
    set<int>visited;
     void dfs(int node,int n){
        visited.insert(node);
        for(int pre:adj[node]){
            if(visited.count(pre)==0) dfs(pre,n);
        }
    }
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        adj.assign(n,vector<int>());
       for(auto& edge:edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
       }
       int res=0;
      
       for(int i=0;i<n;i++){
          if(visited.count(i)>0)
            continue; 
        else {
            dfs(i,n);
            res++;
            }    }
       return res;
    }
};

File: Data Structures & Algorithms/valid-tree/submission-2.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges)!= n-1:
            return False
        adj=[[]*n for _ in range(n)]
        for u,v in edges:
            adj[u].append(v)
            adj[v].append(u)
        visited=set()
        stack=[0]
        while stack :
            curr=stack.pop()
            visited.add(curr)
            for nei in adj[curr]:
                if nei not in visited:
                    stack.append(nei)

        return len(visited)==n